rfc/plugins.md

```diff
@@ -0,0 +1,206 @@
+
+In general, OSS plugins should focus on doing a specific task. Most
+Ember developers with common deployment targets will compose multiple
+plugins to fine tune their deployment process. Developers with very
+custom needs might create a single private plugin that implements all
+aspects of their deployment process within the structure provided by
+ember-cli-deploy.
+
+Because plugins are implemented via addons, they may be included via
+node_module dependencies or via in-repo-addons.
+
+### Identifying plugins
+
+Plugins are ember-cli addons. They will also have the keyword
+`ember-cli-deploy-plugin`. An example `package.json` for a plugin:
+
+```
+{
+  "name": "ember-cli-deploy-example-plugin",
+  "version": "0.4.0",
+  // ...
+  "devDependencies": {
+    "ember-cli": "0.2.0",
+    "ember-cli-deploy": "0.4.1"
+    // ...
+  },
+  "keywords": [
+    "ember-addon",
+    "ember-cli-deploy-plugin"
+  ]
+}
+```
+
+By default, any plugin that is installed will be loaded, similar to
+ember-cli addons. The plugin order is determined by how ember-cli orders
+addons (based on `before:`/`after:` properties). To override this, see
+Advanced Plugin Configuration and Ordering below.
+
+### Plugins are provided by ember-cli addons
+
+Plugins are ember-cli addons which implement an `createDeployPlugin()` method
+to return an object which implements one or more methods that are called by
+ember-cli-deploy. For example:
+
+```javascript
+module.exports = {
+  name: 'ember-cli-deploy-example-plugin',
+  createDeployPlugin: function(options){
+    return {
+      name: options.name,
+      willDeploy: function:(deployment){
+        // do something during the willDeploy phase of the pipeline
+      },
+      didDeploy: function:(deployment){
+        // do something during the didDeploy phase of the pipeline
+      },
+      // etc, see hooks section for a complete list of methods that
+      // may be implemented by a plugin
+    }
+  }
+}
+```
+
+This approach limits the risk of name conflicts at the top-level of the addon.
+It also allows for the plugin author to consult the addon instance during
+creation of the plugin object to make any contextual decisions necessary.
+Finally, it is agnostic with respect to the type of object the plugin is.
+It may be a POJO, a subclass of CoreObject, or maybe an ES6 class.
+
+The `options` argument passed to `createDeployPlugin` will have a `name`
+property. Usually, the `name` will be the plugin name sans the `ember-cli-deploy-`
+prefix, unless a name has been specified as described in Advanced Plugin
+Configuration below.
+
+### The `deployment` object
+
+For each high-level ember-cli-deploy operation, a `deployment` object is created.
+This object is passed to each hook that is invoked on the plugins. It has a number
+of properties that may be of use to a plugin:
+
+Property | file | info
+--- | --- | ---
+`ui` | - | The ember-cli UI object that can be used to write to stdout.
```

Comment:
@lukemelia What are your thoughts about supplying the `analytics` and `project` objects in this `deployment` object as well?  The Ember CLI BuildTask  requires both of them I believe.  The `project` object is present on the `config` property however this doesn't quite feel right to me.

Thoughts?